DigitalEcosystem/data_splitting/bg_extra_descriptors.py

```diff
@@ -0,0 +1,327 @@
+
+
+# # Data Cleaning
+
+# In[]:
+
+
+df = pd.read_pickle("../raw_data/2d_mat_dataset_raw.pkl")
+total = len(df)
+print(f"Starting with {total} entries. Includes top-down and bottom-up.")
+      
+# Throw out systems that are predicted to decompose
+df = df[df["decomposition_energy (eV/atom)"] == 0]
+print(f"Discarding {total-len(df)} entries predicted to decompose (Decomp energy != 0). Total is now {len(df)}.")
+total = len(df)
+
+# Remove systems containing f-block / synthetic elements
+fblock_and_synth = ["La", "Ce", "Pr", "Nd", "Pm", "Sm", "Eu", "Gd", "Tb", "Dy", "Ho", "Er", "Tm", "Yb", "Lu",
+                    "Ac", "Th", "Pa", "U", "Np", "Pu", "Am", "Cm", "Bk", "Cf", "Es", "Fm", "Md", "No", "Lr",
+                    "Rf", "Db", "Sg", "Bh", "Hs", "Mt", "Ds", "Tg", "Cn", "Nh", "Fl", "Mc", "Mc", "Lv", "Ts", "Og"]
+df = df[df["atoms_object (unitless)"].apply(
+           lambda atoms: all(symbol not in fblock_and_synth for symbol in atoms.get_chemical_symbols()))
+       ]
+print(f"Discarding {total - len(df)} entries in the f-block, or synthetics. Total is now {len(df)}.")
+total = len(df)
+
+# Discard metallic systems
+df = df[df["bandgap (eV)"] > 0]
+print(f"Discarding {total - len(df)} entries that are metallic. Total is now {len(df)}.")
+total = len(df)
+
+# To keep the methodology consistent, discard systems where a U-correction was applied
+df = df[df["is_hubbard (unitless)"] == False]
+print(f"Discarding {total - len(df)} entries that have U-corrections applied (metal oxides/fluorides). Total is now {len(df)}.")
+total = len(df)
+
+# Throw out systems based on exfoliation energy
+min_exfol = 0.0
+max_exfol = 0.2
+df = df[df["exfoliation_energy_per_atom (eV/atom)"].apply(lambda exfol: min_exfol < exfol <= max_exfol)]
+print(f"Discarding {total - len(df)} entries outside {np.round(min_exfol,3)} < exfoliation energy <= {np.round(max_exfol,3)}. "
+      f"Total is now {len(df)}.")
+
+
+# # Featurization
+
+# In[]:
+
+
+structures = df['atoms_object (unitless)'].apply(pymatgen.io.ase.AseAtomsAdaptor.get_structure)
+df['ox_struct'] = structures.apply(lambda i: i.copy())
+
+# struct.add_oxidation_state_by_guess() modifies the structure in-place
+df.ox_struct.swifter.apply(lambda struct: struct.add_oxidation_state_by_guess())
+
+
+# In[]:
+
+
+struct_features = MultipleFeaturizer([
+    EwaldEnergy(),
+    StructuralComplexity(),
+    GlobalSymmetryFeatures('n_symmetry_ops'),
+])
+df[struct_features.feature_labels()] = struct_features.featurize_many(df.ox_struct).copy()
+
+
+# In[]:
+
+
+desc = GlobalInstabilityIndex()
+
+def maybe_global_instability(struct):
+    try:
+        return desc.featurize(struct)[0]
+    except:
+        return None
+
+with warnings.catch_warnings():
+    warnings.simplefilter('ignore')
+    df['global_instability'] = df.ox_struct.swifter.apply(maybe_global_instability).copy()
+
+
+# In[]:
+
+
+neighbor_finder = JmolNN()
+
+def average_bond_length(structure, featurizer = AverageBondLength(neighbor_finder)):
```

Comment:
We should start building out content like this in a separate file ("features.py"?) and import from there